### PR TITLE
Fixes blocking RingBuffer ReadManyOperation

### PR DESCRIPTION
**Pull request by mustafaiman**
ReadManyOperation(async) blocks when there is no item in the ringbuffer. Then the thread is parked. When an add operation adds new items, it attempts to unblock waiting read operations. Before unparking the op, the system calls the op's shouldWait method again. shouldWait throws because now sequence is stale (AddAll operation added too many items and overwritten the first one). Since it throws during unparking, the exception is not sent to the requesting party. It is logged and dropped instead. This PR catches StaleSequnceException within shouldWait and lets the operation unblock. The unblocked operation, then returns exception to the calling party.

Fixes https://github.com/hazelcast/hazelcast/issues/13705

### CODE REVIEW

File: hazelcast/src/test/java/com/hazelcast/ringbuffer/impl/operations/ReadManyOperationTest.java [hunk 1]
```diff
@@ -103,15 +108,17 @@ public void whenOneAfterTail() throws Exception {
...
+    @Test
     public void whenTooFarAfterTail() throws Exception {
         ringbuffer.add("tail");
 
         ReadManyOperation op = getReadManyOperation(ringbuffer.tailSequence() + 2, 1, 1, null);
         op.setNodeEngine(nodeEngine);
 
         // since there is an item, we don't need to wait
-        op.shouldWait();
+        assertFalse(op.shouldWait());
```
Comment (mustafaiman) [inline hazelcast/src/test/java/com/hazelcast/ringbuffer/impl/operations/ReadManyOperationTest.java]: Before, `op.shouldWait` threw directly. Now `shouldWait` returns false and `beforeRun` throws. This is an implementation detail, therefore I changed the test to comply the new implementation.

File: hazelcast-client/src/test/java/com/hazelcast/client/ringbuffer/RingbufferTest.java [hunk 2]
```diff
@@ -70,6 +74,13 @@ public void tearDown() {
         hazelcastFactory.terminateAll();
     }
 
+    @Test(expected = ExecutionException.class)
+    public void readManyAsync_whenHitsStale_shouldNotBeBlocked() throws Exception {
+        ICompletableFuture<ReadResultSet<String>> f = clientRingbuffer.readManyAsync(0, 1, 10, null);
+        serverRingbuffer.addAllAsync(asList("0", "1", "2", "3", "4", "5", "6", "7", "8", "9", "10"), OVERWRITE);
+        f.get();
```
Comment (vbekiaris) [inline hazelcast-client/src/test/java/com/hazelcast/client/ringbuffer/RingbufferTest.java]: `f.get()` will throw `ExecutionException` wrapping any other kind of exception. Maybe best to match the root cause with an `ExpectedException` rule?
```
expected.expect(new RootCauseMatcher(StaleSequenceException.class));
f.get();
```

**Comment by mustafaiman** [comment]
@vbekiaris `RingBuffer`s `readOne` is only a sync operation. Notifier scenario does not apply to it. However, I think it is still a good idea to make it follow the same pattern(not throw in `shouldWait`, throw in `beforeRun` instead). I'll make the change.

Regarding `EventJournalReadOperation`, it does not offer an async read method as far as I see. Again the scenario does not apply.

File: hazelcast/src/main/java/com/hazelcast/ringbuffer/impl/operations/ReadManyOperation.java [hunk 3]
```diff
@@ -79,11 +79,15 @@ public boolean shouldWait() {
             return false;
         }
 
-        if (ringbuffer.shouldWait(sequence)) {
+        if (sequence > ringbuffer.tailSequence() + 1
```
Comment (mmedenjak) [inline hazelcast/src/main/java/com/hazelcast/ringbuffer/impl/operations/ReadManyOperation.java]: Nice fix. I'd suggest just one improvement - let's add two methods:
- `RingbufferContainer.isStaleSequence(long sequence)`
-  `RingbufferContainer.isTooLargeSequence(long sequence)`
We can then reuse these methods in the operations and in `RingbufferContainer#checkBlockableReadSequence`.

**Comment by mmedenjak** [comment]
Regarding the single item reading methods, they are also blocking in the sense they can request the next item which has not yet been added to the RB. I guess the same issue can happen there - if the `addAll` method adds enough items to overwrite even the next (not yet written) sequence, even the read-one operations (the `ReadOneOperation`, `MapEventJournalReadOperation` and `CacheEventJournalReadOperation`) can fail.

**Comment by mustafaiman** [comment]
@mmedenjak I added readMany test for EventJournal. It does not fail either cache or map. As far as I understand, event journal add operations are done one by one even when map or cache's putAll causes it. Therefore, a single put cannot overwrite an already waiting read call. So it does not cause blocking. This is what I understood.

**Comment by mustafaiman** [comment]
run-lab-run

**Pull request merged**